crates/alloy-dyn-contract/src/call.rs

```diff
@@ -0,0 +1,187 @@
+    pub fn legacy(self) -> Self {
+        todo!()
+    }
+
+    /// Sets the `gas` field in the transaction to the provided value
+    pub const fn gas(mut self, gas: U256) -> Self {
+        self.request.gas = Some(gas);
+        self
+    }
+
+    /// Sets the `gas_price` field in the transaction to the provided value
+    /// If the internal transaction is an EIP-1559 one, then it sets both
+    /// `max_fee_per_gas` and `max_priority_fee_per_gas` to the same value
+    pub const fn gas_price(mut self, gas_price: U256) -> Self {
+        // todo: differentiate between eip-1559 and legacy once we have typedtx
+        self.request.gas_price = Some(gas_price);
+        self.request.max_fee_per_gas = Some(gas_price);
+        self.request.max_priority_fee_per_gas = Some(gas_price);
+        self
+    }
+
+    /// Sets the `value` field in the transaction to the provided value
+    pub const fn value(mut self, value: U256) -> Self {
+        self.request.value = Some(value);
+        self
+    }
+
+    /// Sets the `block` field for sending the tx to the chain
+    pub const fn block(mut self, block: BlockId) -> Self {
+        self.block = Some(block);
+        self
+    }
+
+    /// Sets the `nonce` field in the transaction to the provided value
+    pub const fn nonce(mut self, nonce: U64) -> Self {
+        self.request.nonce = Some(nonce);
+        self
+    }
+
+    /// Sets the [state override set](https://geth.ethereum.org/docs/rpc/ns-eth#3-object---state-override-set).
+    ///
+    /// # Note
+    ///
+    /// Not all client implementations will support this as a parameter to `eth_call`.
+    pub fn state(mut self, state: StateOverride) -> Self {
+        self.state = Some(state);
+        self
+    }
+
+    /// Returns the underlying transaction's ABI encoded data
+    pub fn calldata(&self) -> Option<Bytes> {
+        self.request.input.data.clone()
+    }
+}
+
+impl<P> CallBuilder<P>
+where
+    P: TempProvider,
+{
+    /// Returns the estimated gas cost for the underlying transaction to be executed
+    pub async fn estimate_gas(&self) -> Result<U256> {
+        self.provider.estimate_gas(self.request.clone(), self.block).await.map_err(Into::into)
+    }
+
+    /// Queries the blockchain via an `eth_call` for the provided transaction.
+    ///
+    /// If executed on a non-state mutating smart contract function (i.e. `view`, `pure`)
+    /// then it will return the raw data from the chain.
+    ///
+    /// If executed on a mutating smart contract function, it will do a "dry run" of the call
+    /// and return the return type of the transaction without mutating the state
+    ///
+    /// # Note
+    ///
+    /// This function _does not_ send a transaction from your account
+    pub async fn call(&self) -> Result<Vec<DynSolValue>> {
+        let bytes = self
+            .provider
+            .call(self.request.clone(), self.block, self.state.clone())
```

Comment:
@DaniPopes  @mattsse is there a world where we can get rid of these clones and use refs under the hood